src/Symfony/Component/Form/Util/FormUtil.php

```diff
@@ -43,30 +43,104 @@ public static function isEmpty($data)
-     * the new array; otherwise, the value from the second array will replace
-     * the one from the first array.
+     * Merges query string or post parameters with uploaded files.
      */
     public static function mergeParamsAndFiles(array $params, array $files): array
     {
-        $isFilesList = array_is_list($files);
+        return self::mergeAtPath($params, $files);
+    }
+
+    private static function mergeAtPath(array $params, array $files, array $path = [])
+    {
+        $paramsValue = self::getValueAtPath($params, $path);
+        $filesValue = self::getValueAtPath($files, $path);
+
+        if (null === $paramsValue) {
+            return $filesValue;
+        }
+
+        if (\is_array($paramsValue) && self::isFileUpload($filesValue)) {
+            return $filesValue; // if the array is a file upload field, it has the precedence
+        }
+
+        if (\is_array($paramsValue) && \is_array($filesValue)) {
+            // if both are lists and both does not contain array, then merge them and return
+            if (array_is_list($paramsValue) && self::doesNotContainNonFileUploadArray($paramsValue) && array_is_list($filesValue) && self::doesNotContainNonFileUploadArray($filesValue)) {
+                return array_merge($paramsValue, $filesValue);
+            }
+
+            // heuristics to preserve order, the bigger array wins
+            if (\count($filesValue) > \count($paramsValue)) {
+                $keys = array_unique(array_merge(array_keys($filesValue), array_keys($paramsValue)));
+            } else {
+                $keys = array_unique(array_merge(array_keys($paramsValue), array_keys($filesValue)));
+            }
+
+            $result = [];
+
+            foreach ($keys as $key) {
+                $subPath = $path;
+                $subPath[] = $key;
 
-        foreach ($params as $key => $value) {
-            if (\is_array($value) && \is_array($files[$key] ?? null)) {
-                $params[$key] = self::mergeParamsAndFiles($value, $files[$key]);
-                unset($files[$key]);
+                $result[$key] = self::mergeAtPath($params, $files, $subPath);
             }
+
+            return $result;
+        }
+
+        if (\is_array($paramsValue)) {
+            return $paramsValue; // params has the precedence
         }
 
-        if (!$isFilesList) {
-            return array_replace($params, $files);
+        if (self::isFileUpload($filesValue)) {
+            return $filesValue; // if the array is a file upload field, it has the precedence
         }
 
-        foreach ($files as $value) {
-            $params[] = $value;
+        return $paramsValue;
+    }
+
+    private static function getValueAtPath(array $params, array $path)
+    {
+        foreach ($path as $key) {
+            if (null === $params = $params[$key] ?? null) {
+                return null;
+            }
         }
 
         return $params;
     }
+
+    /**
+     * @param UploadedFile|array $value
```

Comment:
this type is wrong. The value can be anything (otherwise, the `is_array` check on line 122 would be totally useless btw)